sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,260 @@
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        /// <remarks>The byte order mark is not included as part of the encoding process.</remarks>
+        public BinaryData(string data)
+        {
+            Bytes = s_encoding.GetBytes(data);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the passed in
+        /// <see cref="ReadOnlyMemory{Byte}"/>.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the passed in
+        /// byte array.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromBytes(byte[] data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance using the passed in string.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromString(string data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The passed in stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromStream(Stream stream) =>
+            FromStreamAsync(stream, false).EnsureCompleted();
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The passed in stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="cancellationToken">An optional<see cref="CancellationToken"/> instance to signal
+        /// the request to cancel the operation.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default) =>
+            await FromStreamAsync(stream, true, cancellationToken).ConfigureAwait(false);
+
+        private static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+            if (stream.CanSeek && stream.Length > int.MaxValue)
```

Comment:
Really this would be Length - Position, no?  If a stream has already consumed all but 10 bytes out of 4 billion, the operation will still succeed.